File: fundiapp/lib/models/invoice.dart
// lib/models/invoice.dart

import 'package:netsaf_fund_app/models/rate_card.dart';

class Invoice {
  final int id;
  final String invoiceNumber;
  final String amount;
  final String status;
  final String statusLabel;
  final DateTime? dueDate;
  final DateTime? paidAt;
  final String? pdfUrl;
  final RateCard? rateCard;
  final Map<String, dynamic>? paymentDetails;
  final String? notes;
  final DateTime createdAt;

  // ✅ ADDED: items and currency
  final List<Map<String, dynamic>>? items;
  final String? currency;

  Invoice({
    required this.id,
    required this.invoiceNumber,
    required this.amount,
    required this.status,
    required this.statusLabel,
    this.dueDate,
    this.paidAt,
    this.pdfUrl,
    this.rateCard,
    this.paymentDetails,
    this.notes,
    required this.createdAt,
    this.items,
    this.currency,
  });

  factory Invoice.fromJson(Map<String, dynamic> json) {
    // Safe parsing helpers
    int parseInt(dynamic value) {
      if (value == null) return 0;
      if (value is int) return value;
      if (value is String) return int.tryParse(value) ?? 0;
      if (value is double) return value.toInt();
      return 0;
    }

    DateTime? parseDateTime(dynamic value) {
      if (value == null) return null;
      if (value is DateTime) return value;
      if (value is String) {
        try {
          return DateTime.parse(value);
        } catch (e) {
          return null;
        }
      }
      return null;
    }

    List<Map<String, dynamic>>? parseItems(dynamic value) {
      if (value == null) return null;
      if (value is List) {
        return value.map((e) => e as Map<String, dynamic>).toList();
      }
      return null;
    }

    return Invoice(
      id: parseInt(json['id']),
      invoiceNumber: json['invoice_number']?.toString() ?? '',
      amount: json['amount']?.toString() ?? '0',
      status: json['status']?.toString() ?? 'pending',
      statusLabel: json['status_label']?.toString() ?? 'Pending',
      dueDate: parseDateTime(json['due_date']),
      paidAt: parseDateTime(json['paid_at']),
      pdfUrl: json['pdf_url']?.toString(),
      rateCard: json['rate_card'] != null
          ? RateCard.fromJson(json['rate_card'] as Map<String, dynamic>)
          : null,
      paymentDetails: json['payment_details'] is Map
          ? json['payment_details'] as Map<String, dynamic>
          : null,
      notes: json['notes']?.toString(),
      createdAt: parseDateTime(json['created_at']) ?? DateTime.now(),
      items: parseItems(json['items']),
      currency: json['currency']?.toString() ?? 'TZS',
    );
  }

  bool get isPaid => status == 'paid';
  bool get isPending => status == 'pending';
  bool get isCancelled => status == 'cancelled';
}

File: nearbyfundi_backend/app/Models/Invoice.php
<?php
// app/Models/Invoice.php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\SoftDeletes;
use Carbon\Carbon;

class Invoice extends Model
{
    use SoftDeletes;

    protected $table = 'invoices';

    protected $fillable = [
        'invoice_number', 'user_id', 'subscription_id', 'rate_card_id',
        'amount', 'currency', 'status', 'due_date', 'paid_at',
        'notes', 'items', 'payment_details', 'pdf_path'
    ];

    protected $casts = [
        'amount' => 'decimal:2',
        'due_date' => 'datetime',
        'paid_at' => 'datetime',
        'items' => 'array',
        'payment_details' => 'array',
    ];

    // Status constants
    const STATUS_PENDING = 'pending';
    const STATUS_PAID = 'paid';
    const STATUS_CANCELLED = 'cancelled';

    public function user()
    {
        return $this->belongsTo(User::class);
    }

    public function subscription()
    {
        return $this->belongsTo(Subscription::class);
    }

    public function rateCard()
    {
        return $this->belongsTo(RateCard::class);
    }

    public static function generateInvoiceNumber(): string
    {
        $prefix = 'INV';
        $year = date('Y');
        $month = date('m');
        $last = self::whereYear('created_at', $year)
                    ->whereMonth('created_at', $month)
                    ->count() + 1;
        return $prefix . $year . $month . str_pad($last, 4, '0', STR_PAD_LEFT);
    }

    public function isPaid(): bool
    {
        return $this->status === self::STATUS_PAID;
    }

    public function isOverdue(): bool
    {
        return $this->status !== self::STATUS_PAID && 
               $this->due_date && 
               $this->due_date->isPast();
    }

    public function getStatusLabelAttribute(): string
    {
        return match($this->status) {
            self::STATUS_PENDING => 'Pending',
            self::STATUS_PAID => 'Paid',
            self::STATUS_CANCELLED => 'Cancelled',
            default => ucfirst($this->status),
        };
    }

    public function getFormattedAmountAttribute()
    {
        return number_format($this->amount, 0, '.', ',') . ' ' . $this->currency;
    }

    public function scopePending($query)
    {
        return $query->where('status', self::STATUS_PENDING);
    }

    public function scopePaid($query)
    {
        return $query->where('status', self::STATUS_PAID);
    }
}